### DISCUSSION TITLE
The DWIM version of magit-diff is never what I want

### DISCUSSION THREAD

**Discussion by dabrahams** [magit/magit#5048] [Ideas]
I'm not even sure what it really does.  The thing I want most often is a diff against `origin/main` and it would be nice not to have to think about "ranges" and then type that ref name out in that case.  I don't have a concrete suggestion yet, and I realize this reads a bit like a complaint, sorry.  I wanted to open a discussion before making a feature requrest.

**Comment by tarsius** [comment]
It shows the most logical diff for the thing *at point*, but that is not what *you* want.

> The thing I want most often is a diff against `origin/main` and it would be nice not to have to think about "ranges" and then type that ref name out in that case.

Sounds like `git diff @{upstream}...`. You can get that by moving to `Unmerged into origin/main` and pressing `RET`. Or `d d`! In that case, that's actually the most logical diff to show for the thing at point.

Or to be able to show that diff instantly, without having to first navigate to the respective section in the status buffer or typing in a range, you could add a dedicated command:

```elisp
(defun my-magit-diff-upstream (&optional args files)
  (interactive (magit-diff-arguments))
  (magit-diff-setup-buffer "@{upstream}..." nil args files 'committed))
```

**Reply by Lenbok** [reply]
Kind of related to this, (and slightly related to the workflow mentioned elsewhere about supporting interactive rebase with `--keep-base`), I'm often in the log view and what I want is to diff the current branch vs the branch point. Cribbing from some stack overflow, I've got:
```elisp
  (defun magit-diff-mbase (&optional args files)
    "Show diff of $(git merge-base master HEAD) to working tree."
    (interactive (list (magit-diff-arguments)))
    (magit-diff-setup-buffer
     (magit-git-string "merge-base" (magit-main-branch) "HEAD") nil args files 'committed))
  (transient-append-suffix 'magit-diff
                             "w" '("m" "Diff to merge-base" magit-diff-mbase))
```
This mostly works, however the resulting diff buffer doesn't seem to have "sections" (`magit-jump-to-diffstat-or-diff` warns about "No diffstat in this buffer".
- How can I fix the diffstat section problem?
- Does that `magit-git-string` call look like the best way to get the branch point?

**Reply by tarsius** [reply]
Look at other `magit-diff-*` commands, they use `(interactive (magit-diff-arguments))`, without the extra `(list ...)`, because `magit-diff-arguments` returns `(ARGS FILES)`. This mistake resulted in `magit-insert-diff` not realizing it has to add `--numstat`, which is required so that `magit-diff-wash-diffstat` can get the filename (`--stat` output cannot be used because that may be abbreviated). So the `file` slot of the diffstat section ends up being nil, which is an invalid state, I did not anticipate when I wrote `magit-jump-to-diffstat-or-diff`, so that ends up showing an inaccurate error message.

> I'm often in the log view and what I want is to diff the current branch vs the branch point

I assume you want that diff while the rebase is already in progress. There is no "current branch" at this time; HEAD is "detached". I assume you understand that, else you wouldn't use HEAD.

By pure luck `magit-main-branch` would sometimes return the correct branch but--read its docstring--it has nothing to do with "the branch that was checked out before HEAD got detached during a rebase", which is what you would want (I believe).

During rebases there's a section titled something like `Rebasing work onto work~7`. The branch mentioned there is the one you would want, so let's look at how we get it there. `H` on that section tells us we have to look at `magit-insert-rebase-sequence`. We can see that there is no ready to use function for this, but we learn that we have to look at the contents of the `head-name` file located either in the `rebase-merge` or `rebase-apply` subdirectory of the gitdir.

**Comment by dabrahams** [comment]
Since this discussion got bumped, one thing I'd like and expect from the DWIM command is that when I have point on a commit or ref, DWIM would give me the diff between the current worktree and that commit or the head of that ref.  That is 99.99% of the time what I want when diffing.

**Reply by tarsius** [reply]
> That is 99.99% of the time what I want when diffing.

For me it is more like 0.01%. Or as I said initially:

> It shows the most logical diff for the thing *at point*, but that is not what *you* want.

In other words, you will have to define your own alternative commands and add bindings to the appropriate keymaps.

**Comment by Lenbok** [comment]
> I assume you want that diff while the rebase is already in progress. There is no "current branch" at this time; HEAD is "detached". I assume you understand that, else you wouldn't use HEAD.
> 
> By pure luck `magit-main-branch` would sometimes return the correct branch but--read its docstring--it has nothing to do with "the branch that was checked out before HEAD got detached during a rebase", which is what you would want (I believe).

Actually, I wasn't thinking about doing the diff while the rebase is in progress, just when outside of a rebase, although that could be useful too. My question was only related to my earlier rebase query by me wanting some commands that conveniently operate with respect to the branch point. Thanks for your answer, the interactive argument fix sorted out my main issue, and I can take a look at getting the rebase head name when I need that.

**Comment by dabrahams** [comment]
> as I said initially:
>
> >  It shows the most logical diff for the thing at point, but that is not what you want.

What threw me off about this statement is that a diff has two endpoints.  _**To me**_ “the most logical diff for the thing at point” means the diff between the thing and my working tree.  I still don't know what DWIM actually does when you do it on a commit, but it sounds like it shows you the same thing as when you just hit RET on that commit, i.e. shows you the changes made by the commit(?) If that's the case, I wonder why you'd want a secondary, slower way to do that than simply hitting RET.

Since I don't know what this thing actually does, I can just continue to not use it, but perhaps for other users you'd consider changing the name to something more descriptive of its intent, since "what I mean" depends very much on who "I" turns out to be.  Looking at the manual, it seems none of the DWIM stuff is described in a way that isn't pretty subjective.  Even if you don't change the names, writing these things down would make the commands more useful for everyone else.
